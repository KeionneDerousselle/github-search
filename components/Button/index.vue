<template>
  <button
    v-bind="$attrs"
    :class="['button', { 'button--loading': loading }]"
    v-on="$listeners">
    <div class="button__content">
      <span
        v-if="$slots.left"
        class="button__content--left">
        <span
          v-if="!$slots.right && loading"
          class="button__spinner" />
        <slot
          v-else
          name="left" />
      </span>

      <span class="button__content--center">
        <slot />
      </span>

      <span
        v-if="($slots.right || (!$slots.left && loading))"
        class="button__content--right">
        <span
          v-if="loading"
          class="button__spinner" />
        <slot
          v-else
          name="right" />
      </span>
    </div>
  </button>
</template>
<script>
export default {
  name: 'Button',

  props: {
    loading: {
      type: Boolean,
      default: false
    }
  }
}
</script>
<style lang="scss">
.button {
  @apply flex appearance-none items-center justify-center align-middle cursor-pointer font-bold text-center no-underline select-none text-base pt-2 pb-2 pl-4 pr-4 border border-gray-400 bg-gray-100 rounded-lg;

  font-family: inherit;
  transition: all 0.3s ease-out;

  &__content {
    @apply mr-0 ml-0;

    transition: margin 0.3s cubic-bezier(0.645, 0.045, 0.355, 1);
  }

  &__content,
  &__content--center,
  &__content--left,
  &__content--right {
    @apply flex items-center justify-center;
  }

  &__content--left,
  &__content--right {
    @apply text-base font-normal;
  }

  &__content--left {
    @apply mr-2;
  }

  &__content--right {
    @apply ml-2;
  }

  &__spinner {
    @apply hidden w-4 h-4 bg-transparent border-solid border-2 border-r-0 rounded-full text-center normal-case text-base not-italic font-normal;

    transition: all 0.3s ease-in-out;
  }

  &:active,
  &:focus {
    @apply outline-none;
  }

  @keyframes spin {
    0% {
      -webkit-transform: rotate(0);
      transform: rotate(0);
      animation-timing-function: cubic-bezier(0.55, 0.055, 0.675, 0.19);
    }

    50% {
      -webkit-transform: rotate(180deg);
      transform: rotate(180deg);
      animation-timing-function: cubic-bezier(0.215, 0.61, 0.355, 1);
    }

    100% {
      -webkit-transform: rotate(360deg);
      transform: rotate(360deg);
    }
  }

  &--loading {
    .button__spinner {
      @apply inline-block;

      animation: spin 1s infinite;
    }
  }

  &:hover:not(:disabled),
  &:focus:not(:disabled) {
    /* transform: translateY(-3px); */
  }

  &:active:not(:disabled) {
    transform: translateY(3px);
  }

  &:disabled {
    @apply cursor-not-allowed opacity-70;
  }
}
</style>
