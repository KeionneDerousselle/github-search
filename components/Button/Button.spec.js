import Button from '.'

describe('Button', () => {
  let wrapper
  let buttonClasses

  describe('when the button props are not provided', () => {
    beforeAll(() => {
      wrapper = shallowPreMocked(Button)
      buttonClasses = wrapper.get('.button').classes()
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should display the button', () => {
      expect(wrapper.get('.button').element).toBeVisible()
    })

    it('should render the center button content', () => {
      expect(wrapper.get('.button .button__content--center').element).toBeVisible()
    })

    it('should not render either the left or right loading spinner by default', () => {
      expect(() => wrapper.get('.button .button__spinner')).toThrow()
    })

    it('should not render the loading button class', () => {
      expect(buttonClasses).not.toContain('button--loading')
    })

    it('should not set the disabled attribute on the button', () => {
      // The disabled attribute should not be defined, but most likely will not equal the value false
      // <button></button>
      expect(wrapper.get('.button').attributes().disabled).toBeFalsy()
    })
  })

  describe('when the button is disabled', () => {
    beforeAll(() => {
      wrapper = shallowPreMocked(Button, {
        attrs: {
          disabled: true
        }
      })
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should not set the disabled attribute on the button', () => {
      // The disabled attribute should be defined, but most likely will not equal the value true
      // <button disabled="disabled"></button>
      expect(wrapper.get('.button').attributes().disabled).toBeTruthy()
    })
  })

  describe('when the button is not disabled', () => {
    beforeAll(() => {
      wrapper = shallowPreMocked(Button, {
        attrs: {
          disabled: false
        }
      })
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should not set the disabled attribute on the button', () => {
      // The disabled attribute should not be defined, but most likely will not equal the value false
      // <button></button>
      expect(wrapper.get('.button').attributes().disabled).toBeFalsy()
    })
  })

  describe('when providing the button with a default slot', () => {
    let contentSlotId

    beforeAll(() => {
      contentSlotId = 'button-content-slot'
      wrapper = shallowPreMocked(Button, {
        slots: {
          default: `<span id="${contentSlotId}"></span>`
        }
      })
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should render the content slot', () => {
      expect(wrapper.get(`.button .button__content .button__content--center #${contentSlotId}`).element).toBeVisible()
    })
  })

  describe('when not providing the button with a default slot', () => {
    let contentSlotId

    beforeAll(() => {
      contentSlotId = 'button-content-slot'
      wrapper = shallowPreMocked(Button)
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should not render the content slot', () => {
      expect(() => wrapper.get(`.button .button__content .button__content--center #${contentSlotId}`)).toThrow()
    })
  })

  describe('when listeners are provided to the button', () => {
    let onClickHandler

    beforeAll(async () => {
      onClickHandler = jest.fn()

      wrapper = mountPreMocked(Button, {
        listeners: {
          click: onClickHandler
        }
      })

      wrapper.trigger('click')
      await wrapper.vm.$nextTick()
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should trigger listener handlers on the appropriate events', () => {
      expect(onClickHandler).toHaveBeenCalled()
    })
  })

  describe('when attributes are provided to the button', () => {
    beforeAll(() => {
      wrapper = mountPreMocked(Button, {
        attrs: {
          type: 'submit'
        }
      })
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should pass any attributes to the button component', () => {
      expect(wrapper.attributes().type).toBe('submit')
    })
  })

  describe('when a right button slot is provided, but the button is not loading', () => {
    let contentSlotId
    let rightSlotId

    beforeAll(() => {
      contentSlotId = 'button-content-slot'
      rightSlotId = 'right-slot'

      wrapper = shallowPreMocked(Button, {
        slots: {
          default: `<span id="${contentSlotId}"></span>`,
          right: `<span id="${rightSlotId}">R</span>`
        },

        propsData: {
          loading: false
        }
      })
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should render the right content slot', () => {
      expect(wrapper.get(`.button .button__content .button__content--right #${rightSlotId}`).element).toBeVisible()
    })

    it('should not render the right loading spinner', () => {
      expect(() => wrapper.get('.button .button__content .button__content--right .button__spinner')).toThrow()
    })

    it('should not render the left slot', () => {
      expect(() => wrapper.get('.button .button__content--left')).toThrow()
    })

    it('should not render the left loading spinner', () => {
      expect(() => wrapper.get('.button .button__content .button__content--left .button__spinner')).toThrow()
    })
  })

  describe('when a right button slot is provided, and the button is loading', () => {
    let contentSlotId
    let rightSlotId

    beforeAll(() => {
      contentSlotId = 'button-content-slot'
      rightSlotId = 'right-slot'

      wrapper = shallowPreMocked(Button, {
        slots: {
          default: `<span id="${contentSlotId}"></span>`,
          right: `<span id="${rightSlotId}">R</span>`
        },

        propsData: {
          loading: true
        }
      })
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should not render the right content slot', () => {
      expect(() => wrapper.get(`.button .button__content .button__content--right #${rightSlotId}`)).toThrow()
    })

    it('should render the right loading spinner', () => {
      expect(wrapper.get('.button .button__content .button__content--right .button__spinner').element).toBeVisible()
    })

    it('should not render the left slot', () => {
      expect(() => wrapper.get('.button .button__content--left')).toThrow()
    })

    it('should not render the left loading spinner', () => {
      expect(() => wrapper.get('.button .button__content .button__content--left .button__spinner')).toThrow()
    })
  })

  describe('when a left button slot is provided, but the button is not loading', () => {
    let contentSlotId
    let leftSlotId

    beforeAll(() => {
      contentSlotId = 'button-content-slot'
      leftSlotId = 'left-slot'

      wrapper = shallowPreMocked(Button, {
        slots: {
          default: `<span id="${contentSlotId}"></span>`,
          left: `<span id="${leftSlotId}">L</span>`
        },

        propsData: {
          loading: false
        }
      })
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should not render the right content slot', () => {
      expect(() => wrapper.get('.button .button__content .button__content--right')).toThrow()
    })

    it('should not render the right loading spinner', () => {
      expect(() => wrapper.get('.button .button__content .button__content--right .button__spinner')).toThrow()
    })

    it('should render the left slot', () => {
      expect(wrapper.get(`.button .button__content--left #${leftSlotId}`).element).toBeVisible()
    })

    it('should not render the left loading spinner', () => {
      expect(() => wrapper.get('.button .button__content .button__content--left .button__spinner')).toThrow()
    })
  })

  describe('when a left button slot is provided, and the button is loading', () => {
    let contentSlotId
    let leftSlotId

    beforeAll(() => {
      contentSlotId = 'button-content-slot'
      leftSlotId = 'left-slot'

      wrapper = shallowPreMocked(Button, {
        slots: {
          default: `<span id="${contentSlotId}"></span>`,
          left: `<span id="${leftSlotId}">L</span>`
        },

        propsData: {
          loading: true
        }
      })
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should not render the right content slot', () => {
      expect(() => wrapper.get('.button .button__content .button__content--right')).toThrow()
    })

    it('should not render the right loading spinner', () => {
      expect(() => wrapper.get('.button .button__content .button__content--right .button__spinner')).toThrow()
    })

    it('should not render the left slot', () => {
      expect(() => wrapper.get(`.button .button__content--left #${leftSlotId}`)).toThrow()
    })

    it('should render the left loading spinner', () => {
      expect(wrapper.get('.button .button__content .button__content--left .button__spinner').element).toBeVisible()
    })
  })

  describe('when a right and left button slot is provided', () => {
    let contentSlotId
    let rightSlotId
    let leftSlotId

    beforeAll(() => {
      contentSlotId = 'button-content-slot'
      rightSlotId = 'right-slot'
      leftSlotId = 'left-slot'

      wrapper = shallowPreMocked(Button, {
        slots: {
          left: `<span id="${leftSlotId}">L</span>`,
          default: `<span id="${contentSlotId}"></span>`,
          right: `<span id="${rightSlotId}">R</span>`
        },

        propsData: {
          loading: false
        }
      })
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should render the right slot', () => {
      expect(wrapper.get(`.button .button__content .button__content--right #${rightSlotId}`).element).toBeVisible()
    })

    it('should not render the right loading spinner', () => {
      expect(() => wrapper.get('.button .button__content .button__content--right .button__spinner')).toThrow()
    })

    it('should render the left slot', () => {
      expect(wrapper.get(`.button .button__content .button__content--left #${leftSlotId}`).element).toBeVisible()
    })

    it('should not render the left loading spinner', () => {
      expect(() => wrapper.get('.button .button__content .button__content--left .button__spinner')).toThrow()
    })
  })

  describe('when a right and left button slot is provided, and the button is loading', () => {
    let contentSlotId
    let rightSlotId
    let leftSlotId

    beforeAll(() => {
      contentSlotId = 'button-content-slot'
      rightSlotId = 'right-slot'
      leftSlotId = 'left-slot'

      wrapper = shallowPreMocked(Button, {
        slots: {
          left: `<span id="${leftSlotId}">L</span>`,
          default: `<span id="${contentSlotId}"></span>`,
          right: `<span id="${rightSlotId}">R</span>`
        },

        propsData: {
          loading: true
        }
      })
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should not render the right slot', () => {
      expect(() => wrapper.get(`.button .button__content .button__content--right #${rightSlotId}`)).toThrow()
    })

    it('should render the right loading spinner', () => {
      expect(wrapper.get('.button .button__content .button__content--right .button__spinner').element).toBeVisible()
    })

    it('should render the left slot', () => {
      expect(wrapper.get(`.button .button__content .button__content--left #${leftSlotId}`).element).toBeVisible()
    })

    it('should not render the left loading spinner', () => {
      expect(() => wrapper.get('.button .button__content .button__content--left .button__spinner')).toThrow()
    })
  })

  describe('when no left or right slot are provided, and the button is loading', () => {
    beforeAll(() => {
      wrapper = shallowPreMocked(Button, {
        propsData: {
          loading: true
        }
      })

      buttonClasses = wrapper.get('.button').classes()
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should set the loading class on the button', () => {
      expect(buttonClasses).toContain('button--loading')
    })

    it('should render the loading spinner on the right side of the button', () => {
      expect(wrapper.get('.button .button__content--right .button__spinner').element).toBeVisible()
    })

    it('should not render the loading spinner on the left side of the button', () => {
      expect(() => wrapper.get('.button .button__content--left .button__spinner')).toThrow()
    })
  })

  describe('when no left or right slot are provided, and the button is not loading', () => {
    beforeAll(() => {
      wrapper = shallowPreMocked(Button, {
        propsData: {
          loading: false
        }
      })

      buttonClasses = wrapper.get('.button').classes()
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should not set the loading class on the button', () => {
      expect(buttonClasses).not.toContain('button--loading')
    })

    it('should not render the loading spinner on the right side of the button', () => {
      expect(() => wrapper.get('.button .button__content--right .button__spinner')).toThrow()
    })

    it('should not render the loading spinner on the left side of the button', () => {
      expect(() => wrapper.get('.button .button__content--left .button__spinner')).toThrow()
    })
  })
})
