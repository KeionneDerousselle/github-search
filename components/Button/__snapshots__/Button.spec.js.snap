// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button when a left button slot is provided, and the button is loading should render as expected 1`] = `
<button class="button button--loading">
  <div class="button__content"><span class="button__content--left"><span class="button__spinner"></span></span> <span class="button__content--center"><span id="button-content-slot"></span></span>
    <!---->
  </div>
</button>
`;

exports[`Button when a left button slot is provided, but the button is not loading should render as expected 1`] = `
<button class="button">
  <div class="button__content"><span class="button__content--left"><span id="left-slot">L</span></span> <span class="button__content--center"><span id="button-content-slot"></span></span>
    <!---->
  </div>
</button>
`;

exports[`Button when a right and left button slot is provided should render as expected 1`] = `
<button class="button">
  <div class="button__content"><span class="button__content--left"><span id="left-slot">L</span></span> <span class="button__content--center"><span id="button-content-slot"></span></span> <span class="button__content--right"><span id="right-slot">R</span></span></div>
</button>
`;

exports[`Button when a right and left button slot is provided, and the button is loading should render as expected 1`] = `
<button class="button button--loading">
  <div class="button__content"><span class="button__content--left"><span id="left-slot">L</span></span> <span class="button__content--center"><span id="button-content-slot"></span></span> <span class="button__content--right"><span class="button__spinner"></span></span></div>
</button>
`;

exports[`Button when a right button slot is provided, and the button is loading should render as expected 1`] = `
<button class="button button--loading">
  <div class="button__content">
    <!----> <span class="button__content--center"><span id="button-content-slot"></span></span> <span class="button__content--right"><span class="button__spinner"></span></span>
  </div>
</button>
`;

exports[`Button when a right button slot is provided, but the button is not loading should render as expected 1`] = `
<button class="button">
  <div class="button__content">
    <!----> <span class="button__content--center"><span id="button-content-slot"></span></span> <span class="button__content--right"><span id="right-slot">R</span></span>
  </div>
</button>
`;

exports[`Button when no left or right slot are provided, and the button is loading should render as expected 1`] = `
<button class="button button--loading">
  <div class="button__content">
    <!----> <span class="button__content--center"></span> <span class="button__content--right"><span class="button__spinner"></span></span>
  </div>
</button>
`;

exports[`Button when no left or right slot are provided, and the button is not loading should render as expected 1`] = `
<button class="button">
  <div class="button__content">
    <!----> <span class="button__content--center"></span>
    <!---->
  </div>
</button>
`;

exports[`Button when not providing the button with a default slot should render as expected 1`] = `
<button class="button">
  <div class="button__content">
    <!----> <span class="button__content--center"></span>
    <!---->
  </div>
</button>
`;

exports[`Button when providing the button with a default slot should render as expected 1`] = `
<button class="button">
  <div class="button__content">
    <!----> <span class="button__content--center"><span id="button-content-slot"></span></span>
    <!---->
  </div>
</button>
`;

exports[`Button when the button is disabled should render as expected 1`] = `
<button disabled="disabled" class="button">
  <div class="button__content">
    <!----> <span class="button__content--center"></span>
    <!---->
  </div>
</button>
`;

exports[`Button when the button is not disabled should render as expected 1`] = `
<button class="button">
  <div class="button__content">
    <!----> <span class="button__content--center"></span>
    <!---->
  </div>
</button>
`;

exports[`Button when the button props are not provided should render as expected 1`] = `
<button class="button">
  <div class="button__content">
    <!----> <span class="button__content--center"></span>
    <!---->
  </div>
</button>
`;
